docs/_extensions/external_content.py

```diff
@@ -0,0 +1,241 @@
+  directory) that should be kept even if they do not exist in the source
+  directory. This option can be useful for auto-generated files in the
+  destination directory.
+"""
+
+import filecmp
+import os
+from pathlib import Path
+import re
+import shutil
+import tempfile
+from typing import Dict, Any, List, Optional
+
+from sphinx.application import Sphinx
+
+__version__ = "0.1.0"
+
+DIRECTIVES = ("figure", "image", "include", "literalinclude")
+"""Default directives for included content."""
+
+EXTERNAL_LINK_URL_PREFIX = (
+    "https://github.com/project-chip/connectedhomeip/blob/master/"
+)
+
+
+def adjust_includes(
+    fname: Path,
+    basepath: Path,
+    encoding: str,
+    link_prefixes: List[str],
+    extensions: List[str],
+    dstpath: Optional[Path] = None,
+) -> None:
+    """Adjust included content paths.
+
+    Args:
+        fname: File to be processed.
+        basepath: Base path to be used to resolve content location.
+        encoding: Sources encoding.
+        link_prefixes: Prefixes of links that are made external.
+        extensions: Filename extensions links to which are not made external.
+        dstpath: Destination path for fname if its path is not the actual destination.
+    """
+
+    if fname.suffix != ".md":
+        return
+
+    dstpath = dstpath or fname.parent
+
+    def _adjust_path(path):
+        # ignore absolute paths, section links, hyperlinks and same folder
+        if path.startswith(("/", "#", "http", "www")) or not "/" in path:
+            return path
+        return Path(os.path.relpath(basepath / path, dstpath)).as_posix()
+
+    def _adjust_links(m):
+        displayed, fpath = m.groups()
+        fpath_adj = _adjust_path(fpath)
+        return f"[{displayed}]({fpath_adj})"
+
+    def _adjust_external(m):
+        displayed, target = m.groups()
+        return f"[{displayed}]({EXTERNAL_LINK_URL_PREFIX}{target})"
+
+    def _adjust_filetype(m):
+        displayed, target, extension = m.groups()
+        if extension.lower() in extensions or target.startswith("http"):
+            return m.group(0)
+
+        return f"[{displayed}]({EXTERNAL_LINK_URL_PREFIX}{target})"
+
+    def _remove_section_links(m):
+        (file_link,) = m.groups()
+        return file_link + ")"
+
+    def _adjust_image_link(m):
+        prefix, fpath, postfix = m.groups()
+        fpath_adj = _adjust_path(fpath)
+        return f"{prefix}{fpath_adj}{postfix}"
+
+    rules = [
+        # Find any links and adjust the path
+        (r"\[([^\[\]]*)\]\s*\((.*)\)", _adjust_links),
+
+        # Find links that lead to an external folder and transform it
+        # into an external link.
+        (
+            r"\[([^\[\]]*)\]\s*\((?:\.\./)*((?:" + "|".join(link_prefixes) + r")[^)]*)\)",
+            _adjust_external,
+        ),
+
+        # Find links that lead to a section within another file and
+        # remove the section part of the link.
```

Comment:
This seems like it significantly reduces the utility of the links....  How many links are affected in practice and how annoying is the resulting behavior?

Reply:
There are currently 13 links to sections within other files. They are to the following locations: 

https://github.com/project-chip/connectedhomeip/blob/master/docs/guides/BUILDING.md#prerequisites
https://github.com/project-chip/connectedhomeip/blob/master/docs/guides/access-control-guide.md#installing-a-group-acl-1
https://github.com/project-chip/connectedhomeip/blob/master/docs/guides/chip_tool_guide.md#interacting-with-zcl-clusters
https://github.com/project-chip/connectedhomeip/blob/master/docs/guides/chip_tool_guide.md#building
https://github.com/project-chip/connectedhomeip/blob/master/docs/guides/esp32/ota.md#generate-chip-ota-image
https://github.com/project-chip/connectedhomeip/blob/master/examples/shell/README.md#chip-shell-command-details

I did discover a bug that also removed the section part of http-links, this has now been fixed.

Reply:
Some of those files that are being linked into are quite large.  Losing the section links will make it extremely hard to find the right content in those files.

What can we do to mitigate that?

Reply:
Is there a way to split the large files? Or add some table of contents at the top of files so that navigating is somewhat easier?

Reply:
Sphinx automatically generates a table of content, this can be seen e.g at the right hand bar here: https://gautesl.github.io/connectedhomeip/docs/guides/chip_tool_guide.html

I have been looking into this, and found the problem for why Sphinx does not recognize the autogenerated anchors on other pages. The `external_content` extension copies all relevant source files to a build folder, and performs some transformations on them as well (so that e.g. links work on both GitHub and in the generated docs). In this process links are transformed so that they are still valid, i.e. pointing to the original location. However, the autogenerated anchors only work for the copied files in the build folder, so links to sections outside of it will not work. This is why links to sections within a page works but not to sections in another page. I can try to come up with a solution to this. Preferably the links should be transformed so they point to the new location of a file, but predicting its new location (or whether it is a part of the build at all) might be hard.

Reply:
@gautesl As long as there is a concrete plan here for making things findable (and making sure that everything being linked to is in fact present as an entry in the TOC might be enough), that's fine.  I just want to make sure we have such a plan.

Reply:
I made some changes to how links in build files are being transformed, so section links should work now. As a consequence, some broken section links (such as https://github.com/project-chip/connectedhomeip/blob/master/docs/guides/chip_tool_guide.md#interacting-with-zcl-clusters listed above) have been changed.

Reply:
@gautesl Thank you!  It's very much appreciated!